ch_pipeline/analysis/flagging.py

```diff
@@ -1702,6 +1702,348 @@ def process(self, ss):
+        stack_new, stack_flag = tools.redefine_stack_index_map(
+            inputmap, data.prod, data.stack, data.reverse_map["stack"]
+        )
+
+        valid_stack = np.flatnonzero(stack_flag)
+        ninvalid = stack_new.size - valid_stack.size
+
+        if ninvalid > 0:
+            stack_new = stack_new[valid_stack]
+            self.log.info(
+                "Could not find appropriate reference inputs for "
+                f"{ninvalid:0.0f} stacked products.  Ignoring these "
+                "products in decorrelated cylinder calculation."
+            )
+
+        t = data.prod[stack_new["prod"]]
+
+        prodstack = t.copy()
+        conj = stack_new["conjugate"]
+        prodstack["input_a"] = np.where(conj, t["input_b"], t["input_a"])
+        prodstack["input_b"] = np.where(conj, t["input_a"], t["input_b"])
+
+        # Calculate baseline distance and polarisation pair
+        index_a = prodstack["input_a"]
+        index_b = prodstack["input_b"]
+
+        bdist = pos[index_a] - pos[index_b]
+        bpol = np.core.defchararray.add(pol[index_a], pol[index_b])
+
+        # Find the grid indices
+        xind, yind, dx, dy = find_grid_indices(bdist)
+
+        # Only use the co-polar, 1-cylinder separation visibilities for this analysis
+        ico = np.flatnonzero((pol[index_a] == pol[index_b]) & (np.abs(xind) == 1))
+
+        # Create an identifier based on the polarisation pair and north-south baseline
+        pol_map = {pstr: pp for pp, pstr in enumerate(np.unique(bpol[ico]))}
+
+        idd = np.zeros((ico.size, 2), dtype=int)
+        idd[:, 0] = [pol_map[bpol[ii]] for ii in ico]
+        idd[:, 1] = yind[ico]
+
+        # Find the unique pol/baselines and the inverse map
+        uidd, index = np.unique(idd, return_inverse=True, axis=0)
+
+        isort = np.argsort(index)
+
+        # Determine boundaries of unique pol/baselines
+        bnd = np.concatenate(
+            ([0], np.flatnonzero(np.diff(index[isort]) > 0) + 1, [index.size])
+        )
+
+        # Ignore any unique pol/baseline that does not have ncyl - 1 redundant copies.
+        # This can happen due to the valid_stack selection.
+        ncopies = bnd[1:] - bnd[:-1]
+        bflag = ncopies == (ncyl - 1)
+        nmeas = np.sum(bflag)
+
+        # Loop over the unique pol/baselines and for each one determine if each of the
+        # cylinders is present in each of the redundant copies
+        pindex = np.zeros((nmeas, ncyl - 1), dtype=int)
+        flag_with = np.zeros((ncyl, nmeas, ncyl - 1), dtype=bool)
+
+        cc = 0
+        for bb in range(nmeas):
+
+            if not bflag[bb]:
+                continue
+
+            pi = ico[isort[bnd[bb] : bnd[bb + 1]]]
+            pindex[cc] = valid_stack[pi]
+
+            flag_with[:, cc, :] = np.char.equal(
+                ucyl[:, np.newaxis], cyl[index_a[pi]][np.newaxis, :]
+            ) | np.char.equal(ucyl[:, np.newaxis], cyl[index_b[pi]][np.newaxis, :])
+
```

Comment:
If just using ints, I think the `np.char.equal` just needs to become `==`
```suggestion
            flag_with[:, cc, :] = (
                (ucyl[:, np.newaxis] == cyl[index_a[pi]][np.newaxis, :]) 
                | (ucyl[:, np.newaxis] == cyl[index_b[pi]][np.newaxis, :])
            )
```